### PR TITLE
Prepare chunks on the fly

### PR DESCRIPTION
**Pull request by Marishka17**
### Motivation and context

Currently, when creating a task, original and compressed chunks are prepared.
This takes a significant amount of time and consumes a lot of disk space.
This PR contains a new way to create a task.
At the time of task creation minimum necessary information about the loaded data is collected.
(mapping of key frames and timestamps for video, dummy chunks for images).
Chunks are created "on the fly" when a request for needed chunk is received. 
Created chunks are stored in a temporary cache of limited size.

### How has this been tested?

Manually
### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [x] I have added tests to cover my changes
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2020 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Comment by nmanovic** [comment]
Create new task dialog:
-	Use color picker icon instead of “Label color”
-	The color picker should show the current color

Color picker dialog:
-	It should have “OK”, “Cancel” and “Reset” buttons (see Gimp screenshot)

Job view:
-	Use “color by” == “label” by default
-	Use color picker dialog implemented for create new task dialog
-	Use color picker icon to choose a color
-	You can choose color only if “color by” equals to “instance” or “group”. Thus “Enter” and “color picker” is enabled only when “color by” != “label”.

File: cvat/apps/engine/media_extractors.py [hunk 1]
```diff
@@ -285,6 +291,19 @@ def save_as_chunk(self, images, chunk_path):
         # and does not decode it to know img size.
         return []
 
+    def save_as_chunk_to_buff(self, images, format_='jpeg'):
+        buff = io.BytesIO()
+
+        with zipfile.ZipFile(buff, 'w') as zip_file:
+            for idx, image in enumerate(images):
+                arcname = '{:06d}.{}'.format(idx, format_)
```
Comment (azhavoro) [inline cvat/apps/engine/media_extractors.py:299]: looks like a non-JPEG image will have the wrong extension

File: cvat/apps/engine/media_extractors.py [hunk 2]
```diff
@@ -341,6 +370,41 @@ def save_as_chunk(self, images, chunk_path):
         output_container.close()
         return [(input_w, input_h)]
 
+    def save_as_chunk_to_buff(self, frames, format_):
```
Comment (azhavoro) [inline cvat/apps/engine/media_extractors.py:373]: I'm not sure the format needs to be customized. Chunk writer can write only one specific format (mp4 or zip). There are `save_as_chunk_to_buff` methods have the same signature, but different meaning of the `format_` argument.

File: cvat/apps/engine/migrations/0029_auto_20200814_2153.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            archive_paths.extend([os.path.join(root, file) for file in files if mimetypes.guess_type(file)[0] in archive_mimes])
+
+    for path in archive_paths:
+        Archive(path).extractall(os.path.dirname(path))
+        os.remove(path)
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('engine', '0028_data_storage_method'),
```
Comment (azhavoro) [inline cvat/apps/engine/migrations/0029_auto_20200814_2153.py:48]: Please squash the migrations to one file

File: cvat/apps/engine/migrations/0029_auto_20200814_2153.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        'application/x-tar',
+        'application/x-tarz',
+        'application/x-tzo',
+        'application/x-xz-compressed-tar',
+        'application/zip',
+    ]
+
+    for data_instance in data_q_set:
+        for root, _, files in os.walk(os.path.join(settings.MEDIA_DATA_ROOT, '{}/raw/'.format(data_instance.id))):
+            archive_paths.extend([os.path.join(root, file) for file in files if mimetypes.guess_type(file)[0] in archive_mimes])
```
Comment (azhavoro) [inline cvat/apps/engine/migrations/0029_auto_20200814_2153.py:39]: Please use `is_archive` function from https://github.com/opencv/cvat/blob/develop/cvat/apps/engine/media_extractors.py#L397

**Comment by nmanovic** [comment]
@Marishka17 , need to fix a couple of codacy issues.

**Comment by nmanovic** [comment]
@Marishka17 , need to resolve conflicts.
@azhavoro , could you please comment or approve the PR?

File: cvat/apps/engine/media_extractors.py [hunk 5]
```diff
@@ -191,7 +186,14 @@ def get_image(self, i):
         return io.BytesIO(self._zip_source.read(self._source_path[i]))
 
     def get_path(self, i):
-        return os.path.join(os.path.dirname(self._zip_source.filename), self._source_path[i])
+        if  self._zip_source.filename:
+            return os.path.join(os.path.dirname(self._zip_source.filename), self._source_path[i])
+        else: #для определения mime_type
```
Comment (azhavoro) [inline cvat/apps/engine/media_extractors.py:191]: Please use English

File: cvat/apps/engine/views.py [hunk 6]
```diff
@@ -27,14 +27,14 @@
 from rest_framework.permissions import SAFE_METHODS, IsAuthenticated
 from rest_framework.renderers import JSONRenderer
 from rest_framework.response import Response
-from sendfile import sendfile
+from sendfile import sendfile#убрать
```
Comment (azhavoro) [inline cvat/apps/engine/views.py:30]: please use English

File: cvat/apps/engine/prepare.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+                    raise Exception('Invalid pts sequences')
+
+                if None not in [frame.dts, frame_dts] and frame.dts <= frame_dts:
+                    self._close_video_container(container)
+                    raise Exception('Invalid dts sequences')
+
+                frame_pts, frame_dts = frame.pts, frame.dts
+        self._close_video_container(container)
+
+# class Frame:
```
Comment (azhavoro) [inline cvat/apps/engine/prepare.py:57]: lets remove this class definition

File: cvat/apps/engine/media_extractors.py (hunk 1 above)
Comment (azhavoro) [reply cvat/apps/engine/media_extractors.py]: Looks like this is still relevant.

**Comment by azhavoro** [comment]
LGTM, but need to fix a couple of minor comments and add some tests.

**Comment by chaoyifei** [comment]
HI,When does this pr  merge

File: cvat/settings/base.py [hunk 8]
```diff
@@ -422,3 +425,17 @@ def generate_ssh_keys():
...
 }
+
+CACHES = {
+   'default' : {
+       'BACKEND' : 'diskcache.DjangoCache',
+       'LOCATION' : CACHE_ROOT,
+       'TIMEOUT' : None,
+       'OPTIONS' : {
+            #'statistics' :True,
+            'size_limit' : 2 ** 40, # 1 тб
```
Comment (nmanovic) [inline cvat/settings/base.py:436]:
```suggestion
            'size_limit' : 2 ** 40, # 1 Tb
```

**Comment by nmanovic** [comment]
I have merged the PR with some minor modifications manually.

**Pull request closed without merging**